core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -949,9 +953,65 @@ public void onSuccess() {
                                         }
 
                                         @Override
-                                        public void onFailure(Throwable t) {
-                                            // TODO: handle catastrophic non-channel failures
-                                            onReplicaFailure(nodeId, exp);
+                                        public void onFailure(Throwable shardFailedError) {
+                                            logger.error("[{}] catastrophic error while failing replica shard [{}] for [{}]", shardFailedError, shardId, shard, exp);
+                                            if (shardFailedError instanceof ShardStateAction.NoLongerPrimaryShardException) {
```

Comment:
My original line of thought was that failing a shard doesn't mean we lose the data but rather we force another copy to be promoted and the data will be rebuilt somewhere.  Revisiting it, I agree that we might not be there yet and I'll be OK with proceeding more slowing- i.e., log and error message and proceed as we did before. Can we add an assertion here so our test infra will at least fail and give us some insight as to what we may have missed?